File: src/ipc/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { ConnectionId } from "../types/connection";

export interface DataEvent {
  id: ConnectionId;
  data: number[];
}

export interface ClosedEvent {
  id: ConnectionId;
  reason: string;
}

export const onSessionData = (handler: (e: DataEvent) => void): Promise<UnlistenFn> =>
  listen<DataEvent>("session:data", (ev) => handler(ev.payload));

export const onConnectionClosed = (h: (e: ClosedEvent) => void): Promise<UnlistenFn> =>
  listen<ClosedEvent>("connection:closed", (ev) => h(ev.payload));

// keep onSessionClosed as an alias listening to the new event name
export const onSessionClosed = onConnectionClosed;

File: src-tauri/src/ipc/events.rs
use crate::session::SessionId;
use serde::Serialize;

pub const EV_DATA: &str = "session:data";
pub const EV_CLOSED: &str = "connection:closed";

#[derive(Serialize, Clone)]
pub struct DataEvent {
    pub id: SessionId,
    pub data: Vec<u8>,
}

#[derive(Serialize, Clone)]
pub struct ClosedEvent {
    pub id: SessionId,
    pub reason: String,
}
